src/ucp/tag/offload/rndv.c

```diff
@@ -0,0 +1,232 @@
+        .md_index = spriv->super.md_index
+    };
+    void *rndv_op;
+
+    rndv_op = uct_ep_tag_rndv_zcopy(ucp_ep_get_fast_lane(req->send.ep,
+                                                         spriv->super.lane),
+                                    req->send.msg_proto.tag,
+                                    &rndv_hdr, sizeof(rndv_hdr), iov, 1, 0,
+                                    &req->send.state.uct_comp);
+
+    if (ucs_unlikely(UCS_PTR_IS_ERR(rndv_op))) {
+        return UCS_PTR_STATUS(rndv_op);
+    }
+
+    req->flags                   |= UCP_REQUEST_FLAG_OFFLOADED;
+    req->send.tag_offload.rndv_op = rndv_op;
+
+    return UCS_OK;
+}
+
+UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_rndv_offload_proto_progress, (self),
+                 uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+
+    return ucp_proto_zcopy_single_progress(
+            req, UCT_MD_MEM_ACCESS_RMA | UCT_MD_MEM_FLAG_HIDE_ERRORS,
+            ucp_tag_rndv_offload_send_func, NULL,
+            ucp_tag_rndv_offload_completion, ucp_tag_rndv_offload_request_init);
+}
+
+ucp_proto_t ucp_tag_rndv_offload_proto = {
+    .name     = "tag/rndv/offload",
+    .desc     = "rendezvous tag offload",
+    .flags    = 0,
+    .init     = ucp_tag_rndv_offload_proto_init,
+    .query    = ucp_proto_single_query,
+    .progress = {ucp_tag_rndv_offload_proto_progress},
+    .abort    = ucp_proto_request_zcopy_abort,
+    .reset    = ucp_proto_request_zcopy_id_reset
+};
+
+/* SW emulation of rndv offload protocol. This protocol sends regular RTS as
+ * offloaded eager message and this RTS is always handled by SW on the receiver.
+ * It can be used when real rndv offload is not supported (e.g. for non-contig
+ * or very large messages) or when it is enabled by protocol selection.
+ */
+static ucs_status_t
+ucp_tag_rndv_offload_sw_proto_init(const ucp_proto_init_params_t *init_params)
+{
+    ucp_worker_h worker        = init_params->worker;
+    ucp_context_h context      = worker->context;
+    ucp_ep_config_t *ep_config = &ucs_array_elem(&worker->ep_config,
+                                                 init_params->ep_cfg_index);
+    ucp_proto_rndv_ctrl_init_params_t params = {
+        .super.super         = *init_params,
+        .super.latency       = 0,
+        .super.overhead      = 40e-9,
+        .super.cfg_thresh    = context->config.ext.rndv_thresh,
+        .super.cfg_priority  = 60,
+        .super.min_length    = ucp_ep_tag_offload_min_rndv_thresh(ep_config),
+        .super.max_length    = SIZE_MAX,
+        .super.min_iov       = 0,
+        .super.min_frag_offs = UCP_PROTO_COMMON_OFFSET_INVALID,
+        .super.max_frag_offs = ucs_offsetof(uct_iface_attr_t,
+                                            cap.tag.eager.max_short),
+        .super.max_iov_offs  = UCP_PROTO_COMMON_OFFSET_INVALID,
+        .super.hdr_size      = 0,
+        .super.send_op       = UCT_EP_OP_EAGER_SHORT,
```

Comment:
why only short? SW TM RTS can also use bcopy..

Reply:
it is not explicitly stated in the API, but uct_ep_tag_rndv_request uses short protocol

Reply:
do we have uct capability for max_rndv_request?

Reply:
no we have just `max_hdr` which is not the same as it represents free space ater RVH/RAVH in offloaded rndv

Reply:
maybe we should add it?

Reply:
we can add that, but probably in a separate PR?

Reply:
yes

Reply:
for now pls add comment